dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/icon/IconTest.java

```diff
@@ -0,0 +1,227 @@
+        String iconKey = "iconKey";
+        iconService.addCustomIcon(
+            new CustomIcon( iconKey, "description", keywords, fileResource.getUid(),
+                currentUserService.getCurrentUser().getUid() ) );
+    }
+
+    @Test
+    void shouldGetAllIconsWhenRequested()
+    {
+        Map<String, DefaultIcon> defaultIconMap = getAllDefaultIcons();
+
+        assertEquals( defaultIconMap.size() + 1, iconService.getIcons().size(),
+            String.format( "Expected to find %d icons, but found %d instead", defaultIconMap.size() + 1,
+                iconService.getIcons().size() ) );
+    }
+
+    @Test
+    void shouldGetDefaultIconWhenKeyBelongsToDefaultIcon()
+        throws NotFoundException
+    {
+        String defaultIconKey = getAllDefaultIcons().keySet().stream().findAny().orElse( null );
+
+        Icon icon = iconService.getIcon( defaultIconKey );
+
+        assertEquals( defaultIconKey, icon.getKey() );
+    }
+
+    @Test
+    void shouldGetAllKeywordsWhenRequested()
+    {
+        Set<String> keywordList = getAllDefaultIcons().values().stream()
+            .map( Icon::getKeywords )
+            .flatMap( Arrays::stream ).collect( Collectors.toSet() );
+
+        assertEquals( keywordList.size() + keywords.length, iconService.getKeywords().size(),
+            String.format( "Expected to find %d icons, but found %d instead", keywordList.size() + keywords.length,
+                iconService.getIcons().size() ) );
+    }
+
+    @Test
+    void shouldGetIconsFilteredByKeywordWhenRequested()
+        throws BadRequestException,
+        NotFoundException
+    {
+        Optional<DefaultIcon> defaultIcon = getAllDefaultIcons().values().stream()
+            .filter( si -> si.getKeywords().length > 0 ).findAny();
+
+        if ( defaultIcon.isEmpty() )
+        {
+            return;
+        }
+
+        String keyword = defaultIcon.get().getKeywords()[0];
+
+        FileResource fileResourceB = createAndPersistFileResource( 'B' );
+        iconService
+            .addCustomIcon(
+                new CustomIcon( "iconKeyB", "description", new String[] { "k4", "k5", "k6" }, fileResourceB.getUid(),
+                    currentUserService.getCurrentUser().getUid() ) );
+        FileResource fileResourceC = createAndPersistFileResource( 'C' );
+        iconService
+            .addCustomIcon(
+                new CustomIcon( "iconKeyC", "description", new String[] { "k6", "k7", "k8" }, fileResourceC.getUid(),
+                    currentUserService.getCurrentUser().getUid() ) );
+        FileResource fileResourceD = createAndPersistFileResource( 'D' );
+        iconService
+            .addCustomIcon( new CustomIcon( "iconKeyD", "description", new String[] { keyword }, fileResourceD.getUid(),
+                currentUserService.getCurrentUser().getUid() ) );
+
+        assertEquals( 1, iconService.getIcons( new String[] { "k4", "k5", "k6" } ).size(),
```

Comment:
Please use assertions like

`assertContainsOnly(List.of(iconB), iconService.getIcons( new String[] { "k4", "k5", "k6" } ))`

they are easier to read and give more context in case the test fails. You'll get the actual icons instead of just sizes. I think you'll need `@EqualsAndHashCode` on the Custom/DefaultIcons.